### ISSUE TITLE
Workflow: Raw recording upload, orthographic transcription, MAUS

### ISSUE THREAD

**Issue by MJochim** [IPS-LMU/emuDB-manager#26]
One very common workflow in creating Emu speech databases is to make audio recordings, transcribe them orthographically and feed the audio-transcription pairs into MAUS for automatic segmentation. Maus offers the option to output the data in emuDB format.

# Limitations

- WebMAUS cannot process a directory structure => we have to upload one session at a time, get one emuDB per session and merge the databases
- txt files containing orthographic transcriptions have to be prepared manually.
  - In some cases (when many recordings share the same carrier sentence and only differ in terms of their target word), this can be (shell-)scripted quite easily, or it might be retrieved from SpeechRecorder (which has the prompts in a structured format).
  - *Problem*: Not possible in every experiment; requires scripting knowledge
  - But in other cases (when the uttered speech cannot be predicted from the prompt), manual transcription work is required and this would be much easier with a tool such as WebTranscribe or the EMU-webApp.
  - *Problem*: As of now, neither WebTranscribe nor the EMU-webApp can straightforwardly be used for this process. A lot of manual setup and conversion is required (and in the case of WebTranscribe, special privileges).
  - Might also be solved with a table or spreadsheet that maps transcriptions and audio recordings. Transcriptions can easily be typed in a tool such as Excel.
  - *Problem*: No standardised way to create the necessary .txt files

# Suggestions

The emuDB Manager should aid in creating Emu speech databases. It could be done this way:

- Allow uploading raw recordings (ie, wav files that are not stored in emuDB format)
- emuDB Manager creates an emuDB containing the recordings but no annotations
- Offer option for transcription
  - Import/export WebTranscribe <-> emuDB Manager
  - Create a level for orthographic transcription and use EMU-webApp to do it (the app would need some improvements for efficient work with labels that long)
  - Import from spreadsheet

And now two alternatives:

- (Automatically) Call emuR to feed the transcribed data into WebMAUS (@NPoe is working on such a feature for emuR)
- Offer .txt files for download, such that the user can manually run MAUS

This would by far reduce the technical skills required to create an emuDB :-).

**Comment by florianschiel** [comment]
> WebMAUS cannot process a directory structure => we have to upload one session at a time, get one emuDB per session and merge the databases

Not sure what you mean with 'directory structure' here. The Web API 'WebMAUS Multiple' can batch process any number of files and automatically creates a emuDB of the results.

> Offer .txt files for download, such that the user can manually run MAUS

One way to send orthographic data stored in a emuDB level (e.g. after manual transcription with EmuWebApp) is to export the emuDB into BPF files (export_BPFCollection()) and then use the Web APIs G2P (to create the KAN tier) followed by WebMAUS General (to create the MAU tier) which then results in a new emuDB with word and phonetic segmentations.
Not the most elegant way to do it, I admit. So, the direct call of MAUS from emuR would be a better way to do it in the future.

**Comment by NPoe** [comment]
> (Automatically) Call emuR to feed the transcribed data into WebMAUS (@NPoe is working on such a feature for emuR)

This feature is pretty much done now, so you could probably start experimenting with it next week. I'll come over later today for a demonstration.

My current workflow:
Create empty emuDB
Import media files (function by @raphywink)
Import transcription files from text file collection (function by me)
or: import transcriptions from a table mapping bundles to texts (to be implemented)
bas_run_maus_from_transcription() (function by me)

I'm also supporting G2P on its own, the Chunker, Pho2Syl and MINNI.

Current issues: RCurl has issues on long uploads or processes. This might be related to a runtime issue we have at the BAS too, and it might take a while until we fix it there. But for shortish recordings it works rather well.

**Comment by MJochim** [comment]
> @florianschiel wrote:
> > WebMAUS cannot process a directory structure => we have to upload one session at a time, get one emuDB per session and merge the databases
>
> Not sure what you mean with 'directory structure' here. The Web API 'WebMAUS Multiple' can batch process any number of files and automatically creates a emuDB of the results.

Sorry, that was not clear enough: What I meant was a "recursive" directory structure that is typical of both SpeechRecorder and emuDB, with a db directory containing subdir A for speaker/session A, subdir B for speaker/session B and so on. At this time (and to the best of my knowledge), WebMaus does not handle both subdirs in one run. (Which is not a big deal, but it is cumbersome once you have so many speakers/sessions.)

As to your other comment: Full ack - we do have a solution, but we should add a more elegant (i.e. user-friendly) one.

**Comment by MJochim** [comment]
@julianpoemp seems to be addressing this with his oh-portal. Would love to hear more about what you are up to, Julian.

**Comment by julianpoemp** [comment]
@MJochim , yes I'm currently working on a web-application that offers a transcription chain: ASR -> OCTRA -> MAUS -> EMU-WebApp. The results of each step can be downloaded in different formats. OCTRA and EMU-WebApp are called in an iframe in the same window.
The core idea is to show the processes in a kind of To-Do-list so that the user has a clear view on what's going on.

![oh-portal01](https://user-images.githubusercontent.com/25926588/38828970-f86283fc-41b7-11e8-80a3-cd749c40690c.png)

<hr/>

After file selection from the last step:

![bildschirmfoto 2018-04-16 um 20 53 27](https://user-images.githubusercontent.com/25926588/38829088-54fac5f2-41b8-11e8-964e-97e2f8a9a178.png)


The main workflow is:

1. User adds an audio file and optionally a transcript (CTM, annotJSON, TextGrid etc.). The files are checked for compability and if more than two channels do exist the channels can be splitted into separate files. Invalid files will be ignored.
2. The user sets options (currently language only). He can deactivate ASR or OCTRA. After this the user clicks on the "upload & process" button.
3. Max. three tasks are processed asynchronously (due technical reasons). While processing the user waits until it's his turn. Optionally he can be notified via desktop notifications so that he can minimize the browser window and do other stuff. For example, he will be notified when he can do the editing with OCTRA.
4. After each step the processing time is recorded. This data can be downloaded for further exploration. The result from each step is converted automatically to other formats like CTM, annotJSON etc..
5. The last step ist opening with EMU-webApp (optional). The results from EMU-webApp can be downloaded there.

I called each combination of a wave file and a transcript "task" and each step "operation". There are five operations (each with inputs and outputs and each has knowledge about the other):
![ohportaloperationen](https://user-images.githubusercontent.com/25926588/38826549-71d82532-41b0-11e8-959a-d03f5cddaa16.jpg)

For the steps "ASR" and "MAUS" I'm using the pipeline services by @florianschiel and @thomaskisler. It works very fine!

The proceedings are saved to the browser (IndexedDB) and the files exist for 24h on the server. I'm currently working on a feature that saves transcript files to the IDB, too. Wave files are too big.

The project is still in development: [https://github.com/IPS-LMU/oh-portal](url)

**Comment by MJochim** [comment]
Wow, thanks for your quick response and the cool development! I am wondering about two things.

Given that OH is for oral history (if I remember correctly), I guess you are designing this for scientists with a limited number of long audio recordings. In phonetics, we often haven the 1-second kind of recordings, but thousands of them. Is the oh-portal intended for use with hundreds or thousands of files? Or is this just not what it is made for? The directory structure you already support will help very much in handling large numbers of recordings. Do you support nested directories, or on the top level only?

Currently, you are storing most of the data within the browser. The results of the "EMU-webApp step" must be saved manually. Do you have any plans to extend this scheme? Like, for example, storing stuff on a dedicated server, or downloading an emuDB containing all files rather than downloading them individually?

I am interested in this because the emuDB Manager - see this thread - could one day have very similar features. But this may no longer be necessary.

**Comment by julianpoemp** [comment]
@MJochim, currently - as I know - it's planned to develop this portal for Oral History people who want to have a "easy-to-use" portal. The Oral History people have to transcribe very long audio files so that this portal is designed to assist these people in transcribing.

> In phonetics, we often haven the 1-second kind of recordings, but thousands of them. Is the oh-portal intended for use with hundreds or thousands of files? Or is this just not what it is made for? 

I understand this problem. Theoretically the portal supports short files, too. But downloading the results is still necessary. I implemented a ferature that let you download all results of a column as a .zip archive. At the moment, you could add a folder with 1000 short audio files, but this would let to a table with a lot of rows. (currently the folder can't be closed and opend. But it's planned now ;) )

> The directory structure you already support will help very much in handling large numbers of recordings. Do you support nested directories, or on the top level only?

The GUI supports only one level, because more levels would lead to an overloaded GUI and decrease usability.  But subfolders could be important, too. That's why subfolders are moved to the toplevel automatically.

> The results of the "EMU-webApp step" must be saved manually. Do you have any plans to extend this scheme? Do you have any plans to extend this scheme? Like, for example, storing stuff on a dedicated server, or downloading an emuDB containing all files rather than downloading them individually?

I think there is currently no plans to extend this scheme (as far as I know). If the EMU-webApp would send it's result via ``window.postMessage()`` the OH-Portal could retrieve this result and do some other steps (OCTRA supports this feature in its URL mode). If EMU-webApp supported this OH-Portal could offer a feature like "Save all results as emuDB". Storing stuff to a dedicated server is currently not planned but possible.
